src/memray/_memray/snapshot.h

```diff
@@ -226,6 +224,88 @@ class HighWatermarkFinder
     IntervalTree<Allocation> d_mmap_intervals;
 };
 
+// Like LocationKey, but considers the native_segment_generation and the
+// allocator to be part of the key. Arguably it's a bug that LocationKey
+// doesn't, for each of these. For now, I'm defining a separate type to avoid
+// scope creep, but we should merge these.
+struct HighWaterMarkLocationKey
+{
+    // Ordered widest field to narrowest, to avoid padding.
+    thread_id_t thread_id;
+    size_t python_frame_id;
+    size_t native_frame_id;
+    size_t native_segment_generation;
+    hooks::Allocator allocator;
+
+    bool operator==(const HighWaterMarkLocationKey& rhs) const;
+};
+
+struct HighWaterMarkLocationKeyHash
+{
+    size_t operator()(const HighWaterMarkLocationKey& p) const
+    {
+        // Keep the fewest bits from the hashes of the fields that vary least.
+        size_t ret = std::hash<hooks::Allocator>{}(p.allocator);
+        ret = (ret << 1) ^ std::hash<size_t>{}(p.native_segment_generation);
+        ret = (ret << 1) ^ std::hash<thread_id_t>{}(p.thread_id);
+        ret = (ret << 1) ^ std::hash<size_t>{}(p.native_frame_id);
+        ret = (ret << 1) ^ std::hash<size_t>{}(p.python_frame_id);
+        return ret;
+    }
+};
+
+class HighWaterMarkAggregator : public AbstractAggregator
+{
+  public:
+    virtual void addAllocation(const Allocation& allocation) override;
+    virtual reduced_snapshot_map_t getSnapshotAllocations(bool merge_threads) override;
+    reduced_snapshot_map_t getHighWaterMarkAllocations(bool merge_threads) const;
+
+    size_t getCurrentHeapSize() const noexcept;
+
+  private:
+    // Number of high water marks found (incremented on the falling edge)
+    uint64_t d_peak_count{};
+    size_t d_heap_size_at_last_peak{};
+    size_t d_current_heap_size{};
+
+    // This class represents allocations observed at some location over time.
+    // When an allocation or deallocation is observed, we first check if a new
+    // peak was discovered after the last time an allocation or deallocation
+    // was performed here. If so, the counters tracking what happened since the
+    // last peak must be merged into allocations_contributed_to_last_known_peak
+    // and bytes_contributed_to_last_known_peak and reset: they didn't
+    // contribute to the previous peak, but they are accounted for in the newly
+    // discovered peak. Finally, we update those counters to account for
+    // allocations or deallocations since the current peak.
+    struct UsageHistory
+    {
+        uint64_t last_known_peak{};
+        size_t allocations_contributed_to_last_known_peak{};
+        size_t bytes_contributed_to_last_known_peak{};
+        // NOTE: Signed, as we could have more deallocations than allocations
+        // since the last peak, or more bytes deallocated than allocated.
+        int64_t count_since_last_peak{};
+        int64_t bytes_since_last_peak{};
```

Comment:
It would use extra space, but I'm tempted to simplify by making this
```cpp
size_t count_deallocated_since_last_peak{};
size_t bytes_deallocated_since_last_peak{};
size_t count_allocated_since_last_peak{};
size_t bytes_allocated_since_last_peak{};
```
Especially because my first attempt at writing this had a bug that only manifested on 32-bit because of arithmetic conversion rules (passing `-some_size_t` to a function accepting `int64_t` wound up passing a negative number on 64-bit, as I expected, but a positive number on 32-bit, and I needed to correct it to `-static_cast<int64_t>(some_size_t)`

Reply:
Narrowing is a problem when considering the integer types in the abstract, but I don't think it's possible to hit the implementation-defined behavior given the values we'll have in our domain. The count can never be greater than 1 or less than `-(SIZE_MAX / PAGE_SIZE)` (we never record more than one allocation at a time, and the largest number of deallocations we can record at one time is 1 per mmap'd region, which are a minimum of 1 page).

It's theoretically possible that the size could be large enough to trigger implementation defined behavior, but that would only come up if someone allocated more than 9 exabytes of memory in a single allocation, which isn't realistic on modern hardware, and probably won't be for a very long time.

Reply:
That said - I do think the code would be easier to read and understand if we stuck with unsigned values for everything, at the cost of trading off some space. If you're happy with the extra memory usage, that works for me.

Reply:
Actually - no, this proposed approach has a bigger problem than the narrowing. If I store the allocations and deallocations separately as unsigned integers, they might overflow. As things are written now, if you do
```cpp
free(malloc(2)); // High water mark
while (true) {
    free(malloc(1));
}
```
then each loop iteration switches `count_since_last_peak` and `bytes_since_last_peak` from 0 to 1 and back.

If we changed this to 4 separate counters, then each loop iteration would instead increment all 4 counters by `1`, and on a 32-bit system we'd overflow on the 4294967296th loop iteration.

Reply:
... OK, hear me out. We _could_ use an unsigned integer instead of a signed integer, but otherwise treat it exactly as we treat it today, allowing it to go "negative" (wrapping around to a large positive number), and count on the fact that `size_t(3) + size_t(-2) == size_t(1)`, even though `size_t(-2)` is a large positive number.

That fixes the narrowing, at the cost of making it much harder to read the structure in a debugger...